File: src/cli/wallets.rs
use anyhow::Result;
use clap::Subcommand;
use tabled::Tabled;

use crate::api::client::ApiClient;
use crate::api::models::{Wallet, WalletBalance, ListResponse};
use crate::cli::auth::require_auth;
use crate::utils::{output, spinner};

#[derive(Subcommand)]
pub enum WalletCommands {
    /// List all wallets
    List {
        #[arg(long)]
        json: bool,
    },
    /// Get a wallet by ID
    Get {
        id: String,
        #[arg(long)]
        json: bool,
    },
    /// Create a new wallet from a JSON file
    Create {
        #[arg(long)]
        json: String,
    },
    /// Top up a wallet
    TopUp {
        /// Wallet ID
        id: String,
        /// JSON body with top-up details
        #[arg(long)]
        json: String,
    },
    /// Get real-time wallet balance
    Balance {
        id: String,
        #[arg(long)]
        json: bool,
    },
}

#[derive(Tabled, serde::Serialize)]
struct WalletRow {
    #[tabled(rename = "ID")]
    id: String,
    #[tabled(rename = "Customer")]
    customer_id: String,
    #[tabled(rename = "Balance")]
    balance: String,
    #[tabled(rename = "Currency")]
    currency: String,
    #[tabled(rename = "Status")]
    status: String,
}

impl From<Wallet> for WalletRow {
    fn from(w: Wallet) -> Self {
        Self {
            id: w.id,
            customer_id: w.customer_id.unwrap_or_default(),
            balance: w.balance.map(|b| format!("{:.2}", b)).unwrap_or_default(),
            currency: w.currency.unwrap_or_default(),
            status: w.wallet_status.map(|s| output::status_badge(&s)).unwrap_or_default(),
        }
    }
}

pub async fn handle(cmd: WalletCommands) -> Result<()> {
    let creds = require_auth()?;
    let client = ApiClient::new(creds)?;

    match cmd {
        WalletCommands::List { json } => {
            let sp = spinner::create_spinner("Fetching wallets...");
            let resp: ListResponse<Wallet> = client.get("/v1/wallets").await?;
            sp.finish_and_clear();
            let rows: Vec<WalletRow> = resp.items.into_iter().map(Into::into).collect();
            println!("{}", output::print_table(&rows, json));
        }
        WalletCommands::Get { id, json } => {
            let sp = spinner::create_spinner("Fetching wallet...");
            let wallet: Wallet = client.get(&format!("/v1/wallets/{}", id)).await?;
            sp.finish_and_clear();
            println!("{}", output::print_detail(&wallet, json));
        }
        WalletCommands::Create { json: file } => {
            let data = std::fs::read_to_string(&file)?;
            let body: serde_json::Value = serde_json::from_str(&data)?;
            let sp = spinner::create_spinner("Creating wallet...");
            let wallet: Wallet = client.post("/v1/wallets", &body).await?;
            sp.finish_and_clear();
            output::success(&format!("Wallet created: {}", wallet.id));
            println!("{}", output::print_detail(&wallet, false));
        }
        WalletCommands::TopUp { id, json: file } => {
            let data = std::fs::read_to_string(&file)?;
            let body: serde_json::Value = serde_json::from_str(&data)?;
            let sp = spinner::create_spinner("Topping up wallet...");
            let resp: serde_json::Value = client.post(&format!("/v1/wallets/{}/top-up", id), &body).await?;
            sp.finish_and_clear();
            output::success(&format!("Wallet {} topped up.", id));
            println!("{}", output::print_detail(&resp, false));
        }
        WalletCommands::Balance { id, json } => {
            let sp = spinner::create_spinner("Fetching balance...");
            let balance: WalletBalance = client.get(&format!("/v1/wallets/{}/balance/real-time", id)).await?;
            sp.finish_and_clear();
            println!("{}", output::print_detail(&balance, json));
        }
    }
    Ok(())
}

File: npm/src/cli/wallets.ts
import * as fs from 'node:fs';
import { ApiClient } from '../api/client.js';
import type { ListResponse, Wallet, WalletBalance } from '../api/models.js';
import * as output from '../utils/output.js';
import { createSpinner } from '../utils/spinner.js';
import { requireAuth } from './auth.js';
import { requiredInput, optionalInput, number } from './prompts.js';
import type { GlobalOpts } from './types.js';

export async function walletsList(globalOpts: GlobalOpts, json: boolean): Promise<void> {
  const creds = requireAuth(globalOpts);
  const client = new ApiClient(creds);
  const sp = createSpinner('Fetching wallets...');
  const resp = (await client.get('/v1/wallets')) as ListResponse<Wallet>;
  sp.stop();
  const rows = (resp.items ?? []).map((w) => ({
    ID: w.id ?? '',
    Customer: w.customer_id ?? '',
    Balance: w.balance !== undefined ? w.balance.toFixed(2) : '',
    Currency: w.currency ?? '',
    Status: w.wallet_status ? output.statusBadge(w.wallet_status) : '',
  }));
  process.stdout.write(output.printTable(rows, json));
}

export async function walletsGet(globalOpts: GlobalOpts, id: string, json: boolean): Promise<void> {
  const creds = requireAuth(globalOpts);
  const client = new ApiClient(creds);
  const sp = createSpinner('Fetching wallet...');
  const wallet = (await client.get(`/v1/wallets/${id}`)) as Wallet;
  sp.stop();
  process.stdout.write(output.printDetail(wallet, json));
}

export async function walletsCreate(globalOpts: GlobalOpts, file?: string): Promise<void> {
  const creds = requireAuth(globalOpts);
  const client = new ApiClient(creds);

  let body: unknown;
  if (file) {
    body = JSON.parse(fs.readFileSync(file, 'utf-8'));
  } else {
    output.info('Create a new wallet (press Enter to use defaults)');
    const customer_id = await requiredInput('Customer ID:');
    const currency = await requiredInput('Currency (e.g. USD):');
    const initial_balance = await number({
      message: 'Initial balance (default: 0):',
      default: 0,
    });
    body = { customer_id, currency, initial_balance: initial_balance ?? 0 };
  }

  const sp = createSpinner('Creating wallet...');
  const wallet = (await client.post('/v1/wallets', body)) as Wallet;
  sp.stop();
  output.success(`Wallet created: ${wallet.id ?? ''}`);
  process.stdout.write(output.printDetail(wallet, false));
}

export async function walletsTopUp(
  globalOpts: GlobalOpts,
  id: string,
  file?: string
): Promise<void> {
  const creds = requireAuth(globalOpts);
  const client = new ApiClient(creds);

  let body: unknown;
  if (file) {
    body = JSON.parse(fs.readFileSync(file, 'utf-8'));
  } else {
    output.info(`Top up wallet ${id}`);
    const credits = await number({
      message: 'Credits to add:',
      validate: (v) => (v !== undefined && v > 0 ? true : 'Must be greater than 0'),
    });
    const description = await optionalInput('Description (optional):');
    body = { credits, ...(description ? { description } : {}) };
  }

  const sp = createSpinner('Topping up wallet...');
  const resp = await client.post(`/v1/wallets/${id}/top-up`, body);
  sp.stop();
  output.success(`Wallet ${id} topped up.`);
  process.stdout.write(output.printDetail(resp, false));
}

export async function walletsBalance(
  globalOpts: GlobalOpts,
  id: string,
  json: boolean
): Promise<void> {
  const creds = requireAuth(globalOpts);
  const client = new ApiClient(creds);
  const sp = createSpinner('Fetching balance...');
  const balance = (await client.get(`/v1/wallets/${id}/balance/real-time`)) as WalletBalance;
  sp.stop();
  process.stdout.write(output.printDetail(balance, json));
}
